libs/isograph-react/src/core/startUpdate.ts

```diff
@@ -1,28 +1,323 @@
+import {
+  callSubscriptions,
+  getParentRecordKey,
+  insertEmptySetIfMissing,
+  type EncounteredIds,
+} from './cache';
+import type { RefetchQueryNormalizationArtifactWrapper } from './entrypoint';
 import {
   stableIdForFragmentReference,
+  type ExtractParameters,
   type ExtractStartUpdate,
+  type ExtractUpdatableData,
   type FragmentReference,
   type UnknownTReadFromStore,
 } from './FragmentReference';
-import type { IsographEnvironment } from './IsographEnvironment';
+import {
+  assertLink,
+  type IsographEnvironment,
+  type Link,
+} from './IsographEnvironment';
+import { readPromise, type PromiseWrapper } from './PromiseWrapper';
+import {
+  readImperativelyLoadedField,
+  readLinkedFieldData,
+  readLoadablySelectedFieldData,
+  readResolverFieldData,
+  readScalarFieldData,
+  type NetworkRequestReaderOptions,
+  type ReadDataResultSuccess,
+} from './read';
+import type { ReaderAst } from './reader';
 
 export function getOrCreateCachedStartUpdate<
   TReadFromStore extends UnknownTReadFromStore,
 >(
   environment: IsographEnvironment,
-  fragmentReference: FragmentReference<TReadFromStore, any>,
+  fragmentReference: FragmentReference<TReadFromStore, unknown>,
   eagerResolverName: string,
+  networkRequestOptions: NetworkRequestReaderOptions,
 ): ExtractStartUpdate<TReadFromStore> {
   return (environment.eagerReaderCache[
     stableIdForFragmentReference(fragmentReference, eagerResolverName)
-  ] ??= createStartUpdate(environment, fragmentReference));
+  ] ??= createStartUpdate(
+    environment,
+    fragmentReference,
+    networkRequestOptions,
+  ));
 }
 
 export function createStartUpdate<TReadFromStore extends UnknownTReadFromStore>(
-  _environment: IsographEnvironment,
-  _fragmentReference: FragmentReference<TReadFromStore, any>,
+  environment: IsographEnvironment,
+  fragmentReference: FragmentReference<TReadFromStore, unknown>,
+  networkRequestOptions: NetworkRequestReaderOptions,
 ): ExtractStartUpdate<TReadFromStore> {
-  return (_updater) => {
-    // TODO start update
+  return (updater) => {
+    let mutableUpdatedIds: EncounteredIds = new Map();
+
+    let data = createUpdatableProxy(
```

Comment:
`createUpdatableProxy` can throw, e.g. if there is missing data. And `updater` can throw, too, if due to a change there is missing data. Or because the user wrote some dumb code.

So, we should probably call `callSubscriptions` in a `finally` block for now, and re-throw the error (?). It's an error that probably won't be caught anywhere! But at least users can try to handle it themselves, if they want to e.g. show a toast.

In the long term, I would prefer:

- we write data to a new store layer (e.g. an optimistic layer), and read through that layer. That simplifies things a bit, e.g. no need for the `lastInvalidated` value.
- we merge that layer into the store when the update completes, and trigger subscriptions
- if we throw, we simply throw away that new optimistic layer.

Reply:
This "new layer -> merge -> trigger subscriptions" is probably a general pattern, as well. We could, for example, write network responses to the store in that way. (Though not obviously clear what benefit that gives us).